### PR TITLE
Gutenberg integration

### PR DESCRIPTION
**Pull request by etoledom**
This PR merges Gutenberg into develop.

- The display of Gutenberg as the posts/pages editor is behind a feature flag and a launch argument. (#10497)
- The only action required for anybody to continue working normally is to run `pod install`.
- If anyone wants to try Gutenberg, is necessary to switch the launch argument ON (the default is off).
- This launch argument will be in place until we implement the in-app switch.

There are still some small details that we will be working on in the remaining time.
The most remarkable is that the `Before you continue...` alert is still shown when the user opens a Gutenberg post with Gutenberg.

**Note:**
There is a bug on React Native that when running on iOS 12, there's a continuous stream of logs referring to `nw_socket_handle_socket_event [] Socket SO_ERROR`. **This will happen only if and after you open the Gutenberg editor**.

This is a known issue by the RN team and I hope they will fix it soon.
(https://github.com/facebook/react-native/issues/21030)

### To test:

@loremattei Please test the release flow to make sure that everything continues working for normal releases.
It would be great to make a Gutenberg Alpha release test running Gutenberg to see if that crash you mentioned before continues happening. If that is the case, a crash log would be super helpful.
- As mentioned by Koke here https://github.com/wordpress-mobile/WordPress-iOS/pull/10520#issuecomment-441011076, you might want to modify the feature flag to test the Alpha release.

@diegoreymendez If you could check the changes made to `AztecPostViewController` would be great. It wasn't more than moving code to share it with `GutenbergViewController`.
If you don't feel confortable with those changes, we still can reset `AztecPostViewController` back.

cc @koke @hypest @pinarol

### CODE REVIEW

**Comment by koke** [comment]
> This is a known issue by the RN team and I hope they will fix it soon

Can you link to the issue in the PR for future reference?

**Comment by etoledom** [comment]
Updated the description with that info. 👍

**Comment by koke** [comment]
Also, for testing the release, they might want to edit `FeatureFlag`, since there's no way to specify a launch argument in release builds

File: WordPress/Classes/ViewRelated/Post/PostEditorNavigationBarManager.swift [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    weak var delegate: PostEditorNavigationBarManagerDelegate?
+
+    // MARK: - Buttons
+
+    /// Dismiss Button
+    ///
+    lazy var closeButton: WPButtonForNavigationBar = {
+        let cancelButton = WPStyleGuide.buttonForBar(with: AztecPostViewController.Assets.closeButtonModalImage, target: self, selector: #selector(closeWasPressed))
+        cancelButton.leftSpacing = AztecPostViewController.Constants.cancelButtonPadding.left
+        cancelButton.rightSpacing = AztecPostViewController.Constants.cancelButtonPadding.right
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostEditorNavigationBarManager.swift]: Considering the changes in this PR I'd say `cancelButtonPadding.left` and `cancelButtonPadding.right` should be moved to this class.

File: WordPress/Classes/ViewRelated/Post/PostEditorNavigationBarManager.swift [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    func reloadPublishButton() {
+        publishButton.setTitle(delegate?.publishButtonText ?? "", for: .normal)
+        publishButton.isEnabled = delegate?.isPublishButtonEnabled ?? true
+    }
+
+    func reloadBlogPickerButton(with title: String, enabled: Bool) {
+
+        let titleText = NSAttributedString(string: title, attributes: [.font: AztecPostViewController.Fonts.blogPicker])
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Post/PostEditorNavigationBarManager.swift]: Same as above... considering the changes in this PR `AztecPostViewController.Fonts.blogPicker` should be moved to this class.

The rationale is that this class shouldn't really care or know about `AztecPostViewController` specifically.

File: WordPress/Classes/ViewRelated/Aztec/Helpers/PostEditorUtil.swift [hunk 3]
```diff
@@ -0,0 +1,430 @@
+import Foundation
+
+typealias PostEditorViewControllerType = UIViewController & PublishablePostEditor
+
+class PostEditorUtil: NSObject {
```
Comment (diegoreymendez) [inline WordPress/Classes/ViewRelated/Aztec/Helpers/PostEditorUtil.swift]: You're doing an awesome cleanup job in this PR.

With this specific class, though, I'm a bit confused about its purpose and scope.

I think having a very generic name ("Util") and a bit of a generic scope (UI, State handling, etc) can make it hard to figure out what goes where moving forward.

I have a sensation by looking at the code that we could split this class into (at least) two classes: one for UI (alerts) and another one for post state handling.

PS: the idea is not to hold this PR back though, but I wanted to share those thoughts with you.

Comment (etoledom) [reply WordPress/Classes/ViewRelated/Aztec/Helpers/PostEditorUtil.swift]: Thanks for checking this out!
The PR regarding this specific section was to enable `Publish` and `Save` functionality to Gutenberg Controller.
I share you concern but with @pinarol we decided to simplify the initial work.
You can see some notes about this here.
https://github.com/wordpress-mobile/WordPress-iOS/pull/10461#pullrequestreview-176444065
TL;DR:
We are aware of the mixing of concerns and we will take a closer look to this and how to solve it in the Beta timeline.

Comment (diegoreymendez) [reply WordPress/Classes/ViewRelated/Aztec/Helpers/PostEditorUtil.swift]: That covers it for me, since it's known and a deliberate choice for now.  Thanks for the clarification.

**Comment by etoledom** [comment]
Thanks @diegoreymendez for your review!

I pushed a commit removing that dependency on `AztecPostViewController` from `PostEditorNavigationBarManager`, moving the needed constants as you suggested.

File: WordPress/WordPress.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -8736,6 +8773,20 @@
...
+			files = (
+			);
+			inputPaths = (
+			);
+			name = "Copy Gutenberg JS";
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "# Print commands before executing them (useful for troubleshooting)\nset -x\nDEST=$CONFIGURATION_BUILD_DIR/$UNLOCALIZED_RESOURCES_FOLDER_PATH\n\nif [[ \"$CONFIGURATION\" = *Debug* && ! \"$PLATFORM_NAME\" == *simulator ]]; then\nIP=$(ipconfig getifaddr en0)\nif [ -z \"$IP\" ]; then\nIP=$(ifconfig | grep 'inet ' | grep -v ' 127.' | cut -d\\   -f2  | awk 'NR==1{print $1}')\nfi\n\necho \"$IP\" > \"$DEST/ip.txt\"\nfi\n\nBUNDLE_FILE=\"$DEST/main.jsbundle\"\ncp ${PODS_ROOT}/Gutenberg/bundle/ios/App.js ${BUNDLE_FILE}\n";
```
Comment (loremattei) [inline WordPress/WordPress.xcodeproj/project.pbxproj]: Can we add quotation marks around the last ${BUNDLE_FILE}? 
The script otherwise fails if there's a space in the path and we have spaces in it for "WordPress Internal" builds.

Comment (etoledom) [reply WordPress/WordPress.xcodeproj/project.pbxproj]: Thank you @loremattei !
I added the Quotes 👍
